File: backend/src/models/NotificationModel.js
import mongoose from "mongoose";

const notificationSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: true,
      trim: true,
    },

    message: {
      type: String,
      required: true,
      trim: true,
    },

    counterId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Counter",
      default: null,
    },

    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      default: null,
    },

    roles: {
      type: [String],
      enum: ["ADMIN", "STAFF", "STUDENT"],
      default: [],
    },

    isGlobal: {
      type: Boolean,
      default: false,
    },

    hiddenFor: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],

    readBy: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],

    type: {
      type: String,
      enum: [
        "PAYMENT_SUCCESS",
        "PAYMENT_RECEIVED",
        "CALLED",
        "SKIPPED",
        "CREATED",
        "INFO",
      ],
      default: "INFO",
    },
  },
  {
    timestamps: true,
  },
);

/* ================= INDEXES ================= */

// fast user notifications
notificationSchema.index({ userId: 1 });

// role-based filtering
notificationSchema.index({ roles: 1 });

// global notifications
notificationSchema.index({ isGlobal: 1 });

// counter-based notifications
notificationSchema.index({ counterId: 1 });

// hide/read optimization
notificationSchema.index({ hiddenFor: 1 });
notificationSchema.index({ readBy: 1 });

// main compound query optimization
notificationSchema.index({
  isGlobal: 1,
  roles: 1,
  counterId: 1,
  userId: 1,
});

export default mongoose.model("Notification", notificationSchema);

File: frontend/lib/models/notification_model.dart
class NotificationModel {
  final String id;
  final String title;
  final String message;

  final String? userId;
  final List<String>? roles;
  final bool isGlobal;
  final List<String>? hiddenFor;

  final bool isRead;
  final DateTime? expiresAt;

  NotificationModel({
    required this.id,
    required this.title,
    required this.message,
    this.userId,
    this.roles,
    this.isGlobal = false,
    this.hiddenFor,
    this.isRead = false,
    this.expiresAt,
  });

  factory NotificationModel.fromJson(Map<String, dynamic> json) {
    return NotificationModel(
      id: json["_id"] ?? "",
      title: json["title"] ?? "",
      message: json["message"] ?? "",
      userId: json["userId"],
      roles: json["roles"] != null ? List<String>.from(json["roles"]) : null,
      isGlobal: json["isGlobal"] ?? false,
      hiddenFor: json["hiddenFor"] != null
          ? List<String>.from(json["hiddenFor"])
          : null,
      isRead: json["isRead"] ?? false,
      expiresAt: json["expiresAt"] != null
          ? DateTime.tryParse(json["expiresAt"])
          : null,
    );
  }
}
